### PR TITLE
SFCC provider

### PR DESCRIPTION
**Pull request by dominiksipowicz**
DEMO: https://salesforce-commerce-cloud-k64k3c49m-vercel-solutions-vtest314.vercel.app/

fixes: #164 

## SalesForce Cloud Commerce Provider
- instead of REST / GraphQL API it's using SDK (not isomorphic)
- `getProducts` use SDK search under the hood
- API endpoint `https://commerce-sfcc-dominiksipowicz.vercel.app/api/catalog/products?search=dress`

## How to test?
- API endpoint `https://commerce-sfcc-dominiksipowicz.vercel.app/api/catalog/products?search=dress`
- storefront: `https://commerce-sfcc-dominiksipowicz.vercel.app/`

## references:
- SDK: https://github.com/SalesforceCommerceCloud/commerce-sdk
- isomorphic SDK (currently not used atm): https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic
- PWA Kit  storefront example: https://pwa-kit.mobify-storefront.com/

## features:
```json
{
  "provider": "sfcc",
  "features": {
    "wishlist": false,
    "cart": false,
    "search": true,
    "customerAuth": false,
    "customCheckout": false
  }
}
```

## Training Material and Documentation:

For a detailed introduction into commerce clouds feature set and data setup please refer to our Training Material and Documentation:

Salesforce Trailhead for B2C Commerce Cloud: https://trailhead.salesforce.com/en/content/learn/trails/cc-overview
Salesforce Trailhead for Development on B2C Commerce Cloud: https://trailhead.salesforce.com/en/content/learn/trails/develop-for-commerce-cloud

B2C Commerce Cloud Documentation:https://documentation.b2c.commercecloud.salesforce.com/DOC1/index.jsp
B2C Commerce Cloud Open Commerce API Doc: https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/usage/OpenCommerceAPI.html?cp=0_15
Developer Center for Commerce Cloud (Commerce APIs specifically): https://developer.salesforce.com/docs/commerce/commerce-api/overview
SLAS Org Admin Setup Guide: https://developer.salesforce.com/docs/commerce/commerce-api/references?meta=slas-admin:Summary

### CODE REVIEW

File: site/tsconfig.json [hunk 1]
```diff
@@ -23,8 +23,8 @@
       "@components/*": ["components/*"],
       "@commerce": ["../packages/commerce/src"],
       "@commerce/*": ["../packages/commerce/src/*"],
-      "@framework": ["../packages/local/src"],
-      "@framework/*": ["../packages/local/src/*"]
+      "@framework": ["../packages/sfcc/src"],
+      "@framework/*": ["../packages/sfcc/src/*"]
```
Comment (dominiksipowicz) [inline site/tsconfig.json:27]:
```suggestion
      "@framework": ["../packages/local/src"],
      "@framework/*": ["../packages/local/src/*"]
```

File: packages/sfcc/src/api/operations/get-all-product-paths.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// import data from '../../data.json'
+
```
Comment (dominiksipowicz) [inline packages/sfcc/src/api/operations/get-all-product-paths.ts:2]:
```suggestion
```

File: site/tsconfig.json (hunk 1 above)
Comment (dominiksipowicz) [reply site/tsconfig.json:27]: @lfades maybe we could add it to `.gitignore` and when doing changes in this file we could "force" add it

File: packages/sfcc/.env.template [hunk 3]
```diff
@@ -0,0 +1 @@
+COMMERCE_PROVIDER=@vercel/commerce-sfcc
```
Comment (lfades) [inline packages/sfcc/.env.template:1]: Is this the only required environment variable?

File: packages/sfcc/src/api/operations/get-product.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    preview?: boolean
+  } = {}): Promise<Product | {} | any> {    
+
+    // TODO: support locale
+    const { sdk, locale } = commerce.getConfig(config) as SFCCConfig
+    const shopperProductsClient = await sdk.getshopperProductsClient()
+    const product = await shopperProductsClient.getProduct({parameters: {id: variables?.slug as string}});
+    const normalizedProduct = normalizeProduct(product)
+  
+    // TODO: add dummy data
```
Comment (lfades) [inline packages/sfcc/src/api/operations/get-product.ts:28]: Why should we add dummy data?

File: packages/sfcc/src/api/utils/get-auth-token.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+// client configuration parameters
+export const clientConfig: ClientConfig = {
+  headers: {
+    authorization: ``
+  },
+  parameters: {
+    clientId: process.env.SFCC_CLIENT_ID || '',
+    organizationId: process.env.SFCC_ORG_ID || '',
+    shortCode: process.env.SFCC_SHORT_CODE || '',
+    siteId: process.env.SFCC_SITE_ID || '',
```
Comment (lfades) [inline packages/sfcc/src/api/utils/get-auth-token.ts:12]: These envs should be part of `.env.example`

**Review by lfades — CHANGES_REQUESTED** [review]
Added some comments!

File: packages/sfcc/src/api/operations/get-product.ts (hunk 4 above)
Comment (dominiksipowicz) [reply packages/sfcc/src/api/operations/get-product.ts]: implemented options and variants

File: packages/sfcc/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    "wishlist": false,
+    "cart": false,
+    "search": true,
+    "customerAuth": false,
+    "customCheckout": false
+  }
+}
+```
+
+## Refrences:
```
Comment (lfades) [inline packages/sfcc/README.md:24]:
```suggestion
## References
```

File: packages/sfcc/src/api/endpoints/catalog/products/get-products.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  let searchTerm = search;
+
+  if (categoryId) {
+    searchTerm = categoryId as string
+  }
+
+  // set default for main category
+  if (!searchTerm) {
+    searchTerm = 'clothing'
+  }
```
Comment (lfades) [inline packages/sfcc/src/api/endpoints/catalog/products/get-products.ts:22]:
```suggestion
  // 'clothing' is our main category default, and a manually set category has priority
  const searchTerm = categoryId ? categoryId as string : search || 'clothing';
```

File: packages/sfcc/src/api/endpoints/catalog/products/get-products.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  });
+  let products = [];
+  let found = false;
+  if (searchResults.total) {
+      found = true;
+      products = normalizeSearchProducts(searchResults.hits) as any[];
+  } else {
+      // TODO: handle this better?
+      console.log("No results for search");
+  }
```
Comment (lfades) [inline packages/sfcc/src/api/endpoints/catalog/products/get-products.ts:40]: Can this check only happen if there's an error?

Comment (dominiksipowicz) [reply packages/sfcc/src/api/endpoints/catalog/products/get-products.ts]: Actually, it's not an error so nothing to handle here. The `found` is set to false from the start so this `else` block is not needed at all.

File: packages/sfcc/src/data.json [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+      "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzU0NDczMjUwMjQ0MjA=",
+      "name": "New Short Sleeve T-Shirt",
+      "vendor": "Next.js",
+      "path": "/new-short-sleeve-t-shirt",
+      "slug": "new-short-sleeve-t-shirt",
+      "price": { "value": 25, "currencyCode": "USD" },
+      "descriptionHtml": "<p><span>Show off your love for Next.js and Vercel with this unique,&nbsp;</span><strong>limited edition</strong><span>&nbsp;t-shirt. This design is part of a limited run, numbered drop at the June 2021 Next.js Conf. It features a unique, handcrafted triangle design. Get it while supplies last – only 200 of these shirts will be made!&nbsp;</span><strong>All proceeds will be donated to charity.</strong></p>",
+      "images": [
+        {
+          "url": "/assets/drop-shirt-0.png",
```
Comment (lfades) [inline packages/sfcc/src/data.json:13]: @dominiksipowicz what is this file for?

File: site/tsconfig.json (hunk 1 above)
Comment (dominiksipowicz) [reply site/tsconfig.json]: reverted

**Review by lfades — APPROVED** [review]

**Pull request merged**